src/backend/distributed/executor/subplan_execution.c

```diff
@@ -56,18 +66,49 @@ ExecuteSubPlans(DistributedPlan *distributedPlan)
 	 */
 	BeginOrContinueCoordinatedTransaction();
 
-	nodeList = ActiveReadableNodeList();
-
 	foreach(subPlanCell, subPlanList)
 	{
 		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
 		PlannedStmt *plannedStmt = subPlan->plan;
 		uint32 subPlanId = subPlan->subPlanId;
+
 		DestReceiver *copyDest = NULL;
 		ParamListInfo params = NULL;
 		EState *estate = NULL;
 
 		char *resultId = GenerateResultId(planId, subPlanId);
+		IntermediateResultsHashEntry *entry = SearchIntermediateResult(
+			intermediateResultsHash, resultId);
+
+		/*
+		 * Write intermediate results to local file only if there is no worker
+		 * node that receives them
+		 */
+		bool writeLocalFile = false;
+
+		nodeList = NIL;
+
+		if (list_length(entry->nodeIdList) > 0)
+		{
+			ListCell *nodeIdCell = NULL;
+			foreach(nodeIdCell, entry->nodeIdList)
+			{
+				WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+
+				nodeList = lappend(nodeList, workerNode);
+				elog(DEBUG4, "%s is send to %s:%d", resultId, workerNode->workerName,
```

Comment:
Can you make this as `DEBUG1`(and the below), otherwise adding regression tests become super hard.

Reply:
done